controllers/kubeadmconfig_controller.go

```diff
@@ -162,14 +162,12 @@ func (r *KubeadmConfigReconciler) Reconcile(req ctrl.Request) (_ ctrl.Result, re
 		}
 	}()
 
-	holdLock := false
-	defer func() {
-		if !holdLock {
-			r.KubeadmInitLock.Unlock(ctx, cluster)
-		}
-	}()
-
 	if !cluster.Status.ControlPlaneInitialized {
+		defer func() {
+			if rerr != nil {
+				r.KubeadmInitLock.Unlock(ctx, cluster)
```

Comment:
There is another Unlock call down beneath this code block that we can delete: https://github.com/kubernetes-sigs/cluster-api-bootstrap-provider-kubeadm/blob/master/controllers/kubeadmconfig_controller.go#L265

Reply:
Without this defer block, the first owner machine of the lock will always hold the lock even if it occurs errors during generating cloud init for bootstrap control plane.

Reply:
Agreed. The defer block must stay. I was suggesting deleting the Unlock call on line 265. Do not remove the defer block! It is necessary!

Reply:
Oh, but rereading the code I'm wrong. The unlock on line 265 has to stay otherwise the lock is held forever.

Reply:
Yes, both of them are necessary.